server/static/dashboard/live.js

```diff
@@ -0,0 +1,244 @@
+    var heading = item.querySelector('.filepath');
+    heading.textContent = filepath;
+    var diff = item.querySelector('.diff code');
+    root.querySelector('.code').appendChild(item);
+
+    getBaseline(file.data.filepath, function(err, baseline) {
+      if (err) {
+        diff.textContent = err;
+        return;
+      }
+      file.subscribe(function() {
+        root.classList.remove('disconnected');
+        updateDiff(diff, baseline, file.data.text);
+        updateMatchDisplay(collabid);
+        updateErrorDisplay(collabid);
+        file.on('op', function(ops) {
+          updateDiff(diff, baseline, file.data.text);
+          updateTypingDisplay(collabid);
+          updateMatchDisplay(collabid);
+          updateErrorDisplay(collabid);
+          scrollToLatestOp(file, ops);
+        });
+      });
+    });
+  });
+}
+
+
+var baselines = {}; // filepath => { ajax, err, baseline }
+// ensure that baseline is only fetched once per file, for all collabs
+function getBaseline(filepath, callback) {
+  baselines[filepath] = baselines[filepath] || {};
+  var data = baselines[filepath];
+  if (!data.ajax) {
+    data.ajax = $.ajax('/baseline/' + project + '/' + filepath).done(function(baseline) {
+      data.baseline = baseline;
+    }).fail(function(req, status, err) {
+      data.err = 'Error fetching baseline: ' + errorToString(req.responseJSON, status, err);
+    });
+  }
+  data.ajax.always(function() {
+    callback(data.err, data.baseline);
+  });
+}
+
+
+function updateDiff(node, baseline, text, op) {
+  if (baseline === undefined || text === undefined) { return; }
+  node.innerHTML = '';
+  var lineNumber = 1;
+  diffLines(baseline, text).forEach(function(part) {
+    if (part.removed) { return; }
+    var elt = document.createElement('div');
+    elt.classList.add('diff-part');
+    elt.classList.add('line-' + lineNumber);
+    elt.appendChild(document.createTextNode(part.value));
+    if (part.added) {
+      elt.classList.add('diff-added');
+    }
+    node.appendChild(elt);
+    lineNumber++;
+  });
+  hljs.highlightBlock(node);
+}
+
+
+function updateMatchDisplay(collabid) {
+  var root = document.querySelector('#collab-' + collabid);
+  if (!root) { return; }
+
+  if (searchregex) {
+    var match = false;
+    collabstofiles[collabid].forEach(function (file) {
+      if (match) { return; }
+      var matchdata = file.data.text.match(searchregex);
+      if (matchdata) {
+        match = true;
+        var linenumber = findLineNumberForIndex(matchdata.input, matchdata.index);
+        scrollToLine(collabid, file.data.filepath, linenumber);
+      }
+    });
+    root.classList.toggle('match', match);
+    countdisplay.innerHTML = document.querySelectorAll('.match').length
+         + '/' + document.querySelectorAll('.collab:not(.disconnected)').length;
+  } else {
+    root.classList.remove('match');
+  }
+}
+
+
+var typingtimeoutids = {};
+
+function updateTypingDisplay(collabid) {
```

Comment:
Apply same debouncing function as for search box? Or same strategy of avoiding IDs map.